src/CommandLine/Core/TokenPartitioner.cs

```diff
@@ -23,11 +24,11 @@ public static
             var scalars = Scalar.Partition(tokenList, typeLookup).Memorize();
             var sequences = Sequence.Partition(tokenList, typeLookup).Memorize();
             var nonOptions = tokenList
-                .Where(t => !switches.Contains(t))
-                .Where(t => !scalars.Contains(t))
-                .Where(t => !sequences.Contains(t)).Memorize();
+                .Where(t => !switches.Contains(t, ReferenceEqualityComparer.Default))
+                .Where(t => !scalars.Contains(t, ReferenceEqualityComparer.Default))
+                .Where(t => !sequences.Contains(t, ReferenceEqualityComparer.Default)).Memorize();
             var values = nonOptions.Where(v => v.IsValue()).Memorize();
-            var errors = nonOptions.Except(values).Memorize();
+            var errors = nonOptions.Except(values, ReferenceEqualityComparer.Default).Cast<Token>().Memorize();

```

Comment:
Are you casting back to Token because the comparer casts everything into Object? What if you made the comparer inherit `IEqualityComparer<Token>` and then called [Object.ReferenceEquals](https://msdn.microsoft.com/en-us/library/system.object.referenceequals(v=vs.110\).aspx) in the implementation rather than `==`? That's a bit more explicit too, and wouldn't need a comment explaining that we're aiming for reference equality on line 16.